cmd/update.go

```diff
@@ -0,0 +1,483 @@
+	default:
+		return []string{arch, "all"}
+	}
+}
+
+func matches(goos, goarch, asset string) bool {
+	if !strings.Contains(asset, goos) {
+		return false
+	}
+	if re, ok := archRE[goarch]; ok {
+		return re.MatchString(asset)
+	}
+	return strings.Contains(asset, goarch)
+}
+
+func findAssetFor(rel *ghRelease, goos, goarch string) (name, url string, err error) {
+	for _, wantArch := range archAliases(goarch) {
+		for _, a := range rel.Assets {
+			if matches(goos, wantArch, strings.ToLower(a.Name)) {
+				return a.Name, a.BrowserDownloadURL, nil
+			}
+		}
+	}
+	names := make([]string, 0, len(rel.Assets))
+	for _, a := range rel.Assets {
+		names = append(names, a.Name)
+	}
+	return "", "", fmt.Errorf("no asset for %s/%s (available: %v)", goos, goarch, names)
+}
+
+func findAsset(rel *ghRelease) (name, url string, err error) {
+	return findAssetFor(rel, runtime.GOOS, runtime.GOARCH)
+}
+
+func findSha256For(rel *ghRelease, binName string) (url string, ok bool) {
+	want := binName + ".sha256"
+	for _, a := range rel.Assets {
+		if a.Name == want {
+			return a.BrowserDownloadURL, true
+		}
+	}
+	return "", false
+}
+
+// doDownload fetches the URL to a temporary file and reports its path.
+func doDownload(ctx context.Context, url string) (string, error) {
+	if !strings.HasPrefix(url, "https://") {
+		return "", errors.New("refusing to download from non-HTTPS URL")
+	}
+
+	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("download failed: %s", res.Status)
+	}
+
+	tmp, err := os.CreateTemp("", "oasis-update-*")
+	if err != nil {
+		return "", err
+	}
+	defer tmp.Close()
+
+	pw := progress.NewProgressWriter(res.ContentLength, progress.WithMessage("Downloading..."))
+	_, err = io.Copy(tmp, io.TeeReader(res.Body, pw))
+	if err != nil {
+		return "", err
+	}
+	return tmp.Name(), nil
+}
+
+func replaceExecutable(dest string, r io.Reader, perm os.FileMode) error {
+	dir := filepath.Dir(dest)
+
+	tmp, err := os.CreateTemp(dir, ".oasis-new-*")
+	if err != nil {
+		return err
+	}
+	tmpPath := tmp.Name()
+
+	cleanup := true
+	defer func() {
+		if cleanup {
+			_ = tmp.Close()
+			_ = os.Remove(tmpPath)
+		}
+	}()
+
+	if _, err = io.Copy(tmp, r); err != nil {
+		return err
+	}
+
+	if perm == 0 {
+		perm = 0o755
+	}
+	if err = tmp.Chmod(perm); err != nil && runtime.GOOS != osWindows {
+		return err
+	}
+
+	if err = tmp.Close(); err != nil {
+		return err
+	}
+
+	if err = os.Rename(tmpPath, dest); err == nil {
+		cleanup = false
+		return nil
+	}
+
+	if runtime.GOOS == osWindows {
+		pending := dest + ".new"
+		if err2 := os.Rename(tmpPath, pending); err2 != nil {
+			return fmt.Errorf("rename failed (%v / %v)", err, err2)
+		}
+		scheduleWindowsMove(pending, dest)
+		cleanup = false
+		fmt.Println("Update staged. It will be applied automatically on next launch.")
+		return nil
+	}
```

Comment:
Did you test this on windows? e.g. inside a virtual machine on your laptop?

Reply:
haven't tested on Windows VM yet, will do today and update

Reply:
looking good now on Windows
<img width="720" height="450" alt="image" src="https://github.com/user-attachments/assets/d80ca6e4-1070-4924-9d23-a27df84cfbd1" />